File: go/src/dcf.go
package dcf

import (
	"context"
	"errors"
	"google.golang.org/grpc"
	pb "github.com/ALH477/DeMoD-Communication-Framework/go/src/proto"
)

type Mode string
const (
	ClientMode Mode = "client"
	ServerMode Mode = "server"
	P2PMode    Mode = "p2p"
	AutoMode   Mode = "auto"
	MasterMode Mode = "master"
)

type DCFError string
var (
	ErrConfigInvalid      = errors.New("invalid config")
	ErrMasterUnreachable  = errors.New("master node unreachable")
	// Add more as per SDK guidelines
)

type DCFClient struct {
	conn   *grpc.ClientConn
	stub   pb.DCFServiceClient
	mode   Mode
	config *Config // From config.go
	// Add redundancy, plugin manager fields
}

func NewDCFClient(configPath string) (*DCFClient, error) {
	cfg, err := LoadConfig(configPath)
	if err != nil {
		return nil, err
	}
	conn, err := grpc.Dial(cfg.Host+":"+string(cfg.Port), grpc.WithInsecure())
	if err != nil {
		return nil, err
	}
	return &DCFClient{
		conn:   conn,
		stub:   pb.NewDCFServiceClient(conn),
		mode:   Mode(cfg.Mode),
		config: cfg,
	}, nil
}

// Methods: Start(), Stop(), SendMessage(data string, recipient string) (string, error), etc.

File: JS/nodejs/src/dcf.js
// nodejs/src/dcf.js
const grpc = require('@grpc/grpc-js');
const protoLoader = require('@grpc/proto-loader');
const WebSocket = require('ws');
const uuid = require('uuid');
const packageDefinition = protoLoader.loadSync('services.proto');
const proto = grpc.loadPackageDefinition(packageDefinition).DCFService;

class DCFNetworking {
  constructor(config) {
    this.config = config; // From config.json (mode, host, port, peers, etc.)
    this.mode = config.mode || 'p2p';
    this.plugins = {}; // For dynamic loading
    this.peers = config.peers || [];
    this.client = null;
    this.server = null;
    this.initTransport();
  }

  initTransport() {
    switch (this.config.transport) {
      case 'gRPC':
        this.client = new proto.DCFService(`${this.config.host}:${this.config.port}`, grpc.credentials.createInsecure());
        break;
      case 'WebSocket':
        this.ws = new WebSocket(`ws://${this.config.host}:${this.config.port}`);
        this.ws.on('message', this.handleMessage.bind(this));
        break;
      // Add UDP/TCP via 'dgram' or 'net' modules for Node.js
    }
  }

  send(data, recipient, callback) {
    const msg = { sender: this.config.node_id, recipient, data, timestamp: Date.now(), sequence: uuid.v4() };
    if (this.config.transport === 'gRPC') {
      this.client.SendMessage(msg, (err, response) => callback(err, response ? response.data : null));
    } else if (this.config.transport === 'WebSocket') {
      this.ws.send(JSON.stringify(msg)); // Fallback serialization; use Protobuf for efficiency
      // Handle response via ws.on('message')
    }
  }

  handleMessage(message) {
    // Deserialize (Protobuf or JSON) and process
    console.log('Received:', message);
  }

  // AUTO Mode: Listen for master commands
  setRole(role) {
    this.mode = role;
    // Re-init based on new role (e.g., start server if 'server')
  }

  // P2P Redundancy: RTT Grouping
  async healthCheck(peer) {
    const start = Date.now();
    // Use gRPC health check or ping
    // Return RTT: Date.now() - start
  }

  loadPlugin(name, path) {
    this.plugins[name] = require(path); // Dynamic import
    // Version check: if (this.plugins[name].getVersion() !== '1.0') throw Error;
  }
}

module.exports = DCFNetworking;
